File: java/src/edu/cwru/eecs293/ttf10/uxb/Hub.java
package edu.cwru.eecs293.ttf10.uxb;

/**
 * Represents a UXB hub, which can have multiple computer-side and peripheral-side connectors.
 *
 * Case Western Reserve University
 * EECS 293  |  Software Craftsmanship  |  2016 Fall Semester
 * Programming Assignment 2  |  Due at beginning of discussion on Wednesday, September 7, 2016
 *
 * @see <a href="Hw2.pdf">https://blackboard.case.edu/bbcswebdav/pid-1379538-dt-content-rid-4276971_1/courses/eecs293_vxl11/Hw2.pdf</a>
 *
 * Theodore Frohlich <ttf10@case.edu>
 */
public class Hub extends AbstractDevice<Hub.Builder> {

    /**
     * Initializes the hub from the given builder.
     * @param builder a builder
     */
    private Hub(Builder builder) {
        super(builder);
    }

    @Override
    public DeviceClass getDeviceClass() {
        return DeviceClass.HUB;
    }


    public static class Builder extends AbstractDevice.Builder<Builder> {

        /**
         * Creates a new builder with the given UXB version, no connectors, and with empty product code and serial number.
         * @param version the UXB version that this device supports
         */
        public Builder(Integer version) {
            super(version);
        }

        /**
         * Initializes the hub with the builder’s version, product code, serial number, and connector list.
         * @return the initialized hub
         * @throws IllegalStateException if the version is null, or the hub has no computer connector, or the hub has no peripheral connector.
         */
        public Hub build() throws IllegalStateException {
            // TODO
            return null;
        }

        /**
         * Returns this builder.
         * @return this builder
         */
        @Override
        protected Builder getThis() {
            return this;
        }

        /**
         *
         * @throws IllegalStateException if and only if the version number is null, or if the hub has no computer connector, or if the hub has no peripheral connector
         */
        @Override
        protected void validate() throws IllegalStateException {
            // TODO
        }

    }  // class Builder

}  // class Hub

File: scala/src/edu/cwru/eecs293/ttf10/uxb/Hub.scala
/*                                                                      *\
**                    Case Western Reserve University                   **
**                                                                      **
**                               EECS 293                               **
**                        Software Craftsmanship                        **
**                          2016 Fall Semester                          **
\*                                                                      */

package edu.cwru.eecs293.ttf10.uxb

import DeviceClass._

/**
  * Represents a UXB hub, which can have multiple computer-side and peripheral-side connectors.
  *
  * @see [[https://blackboard.case.edu/bbcswebdav/pid-1381847-dt-content-rid-4318401_1/xid-4318401_1 Hw3.pdf]]
  * <br> [[https://blackboard.case.edu/bbcswebdav/pid-1379538-dt-content-rid-4287477_1/xid-4287477_1 Hw2.pdf]]
  * @since Programming Assignment 2
  * @author Ted Frohlich
  */
class Hub(private val builder: Hub.Builder) extends AbstractDevice(builder) {
  
  override def getDeviceClass: DeviceClass = DeviceClass.HUB
  
  @throws[NullPointerException]
  @throws[IllegalArgumentException]
  def recv(message: StringMessage, connector: Connector) {
    validateRecv(message, connector)
    println("[Log] >>  " + "recv not yet supported")
  }
  
  @throws[NullPointerException]
  @throws[IllegalArgumentException]
  def recv(message: BinaryMessage, connector: Connector) {
    validateRecv(message, connector)
    println("[Log] >>  " + "recv not yet supported")
  }
  
}


object Hub {
  
  /**
    * Creates a new builder with the given UXB version, no connectors, and with empty product code and serial number.
    *
    * @param version the UXB version that this device supports
    */
  class Builder(override protected val version: Int) extends AbstractDevice.Builder[Builder](version) {
  
    protected def getThis = this
  
    /**
      * Initializes the hub with the builder’s version, product code, serial number, and connector list.
      *
      * @return the initialized hub
      * @throws IllegalStateException if the version is null, or the hub has no computer connector, or the hub has no peripheral connector.
      */
    @throws[IllegalStateException]
    def build(): Hub = {
      validate()
      new Hub(this)
    }
  
    /**
      * Validates this builder.
      *
      * @throws IllegalStateException if and only if the version number is null, or if the hub has no computer connector, or if the hub has no peripheral connector
      */
    @throws[IllegalStateException]
    override protected def validate() {
      // Check if the version number is null
      try {
        super.validate()
      } catch {
        case e: NullPointerException => throw new IllegalStateException(e)
      }
      // Check if the hub has no computer connector
      if (!getConnectors.contains(Connector.Type.COMPUTER)) {
        throw new IllegalStateException("Validation failed: hub has no computer connector.")
      }
      // Check if the hub has no peripheral connector
      if (!getConnectors.contains(Connector.Type.PERIPHERAL)) {
        throw new IllegalStateException("Validation failed: hub has no peripheral connector.")
      }
    }
  
  }
  
}
